File: implementations/opshin/src/lp_minting_policy.py
#!opshin
from opshin.prelude import *


def has_nft(
    pool_nft_cs: PolicyId, pool_nft_tn: TokenName, tx_in_info: TxInInfo
) -> bool:
    return tx_in_info.resolved.value[pool_nft_cs][pool_nft_tn] == 1


def validator(
    pool_nft_cs_raw: Anything, redeemer: Anything, context: ScriptContext
) -> None:
    pool_nft_cs: PolicyId = pool_nft_cs_raw
    purpose = context.purpose
    if isinstance(purpose, Minting):
        own_symbol = purpose.policy_id
    else:
        assert False, "Invalid purpose"

    minted_names = context.tx_info.mint[own_symbol].keys()
    assert all(
        [
            any(
                [
                    has_nft(pool_nft_cs, minted_lp_name, inp)
                    for inp in context.tx_info.inputs
                ]
            )
            for minted_lp_name in minted_names
        ]
    )

File: implementations/plutus-tx/lib/UplcBenchmark/LpMintingPolicy.hs
{-# LANGUAGE NoImplicitPrelude #-}
{-# OPTIONS_GHC -Wno-incomplete-uni-patterns #-}

module UplcBenchmark.LpMintingPolicy (lpMintingPolicy) where

import PlutusLedgerApi.V1.Value (valueOf)
import PlutusLedgerApi.V3 (
  BuiltinData,
  CurrencySymbol,
  ScriptContext (scriptContextScriptInfo, scriptContextTxInfo),
  ScriptInfo (MintingScript),
  TokenName,
  TxInInfo (txInInfoResolved),
  TxInfo (txInfoInputs, txInfoMint),
  TxOut (txOutValue),
  UnsafeFromData (unsafeFromBuiltinData),
  Value (Value),
 )
import PlutusLedgerApi.V3.MintValue (MintValue (UnsafeMintValue))
import PlutusTx.Prelude (
  Bool,
  BuiltinUnit,
  Integer,
  all,
  any,
  check,
  traceIfFalse,
  ($),
  (==),
 )
import UplcBenchmark.Utils (getOwnMint)

{-# INLINE lpMintingPolicy #-}
lpMintingPolicy :: BuiltinData -> BuiltinData -> BuiltinUnit
lpMintingPolicy rawPoolNftCs rawCtx =
  let
    !ctx :: ScriptContext = unsafeFromBuiltinData rawCtx
    !poolNftCs :: CurrencySymbol = unsafeFromBuiltinData rawPoolNftCs
    MintingScript !ownSymbol = scriptContextScriptInfo ctx

    inputHasNft :: TokenName -> TxInInfo -> Bool
    inputHasNft poolNftTn txIn =
      let !txOut = txInInfoResolved txIn
       in valueOf (txOutValue txOut) poolNftCs poolNftTn == 1

    !txInfo = scriptContextTxInfo ctx
    !inputs = txInfoInputs txInfo
    UnsafeMintValue mint = txInfoMint txInfo

    isValidMint :: (TokenName, Integer) -> Bool
    isValidMint (mintedLpName, _mintedLpAmount) =
      any (inputHasNft mintedLpName) inputs

    !ownMint = getOwnMint ownSymbol (Value mint)

    validMint :: Bool
    !validMint = all isValidMint ownMint
   in
    check $ traceIfFalse "NFT corresponding to LP must be spent" validMint
